### PR TITLE
Refactor `assignScriptRedeemers`

### PR DESCRIPTION
**Pull request by Anviking**
- [x] Unify the `Babbage-` and `Conway-` variants by instead requiring the right constraints.

Depends and targets #4014 

### Issue Number

ADP-3081 - wanted to refactor it before I moved it

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 1]
```diff
@@ -986,36 +966,17 @@ assignScriptRedeemers pparams timeTranslation utxo redeemers tx =
...
-
-    assignNullRedeemersConway
-        :: ConwayTx
-        -> Either ErrAssignRedeemers (Map Alonzo.RdmrPtr Redeemer, ConwayTx)
-    assignNullRedeemersConway conwayTx = do
+    assignNullRedeemers ledgerTx = do
         (indexedRedeemers, nullRedeemers) <- fmap unzip $ forM redeemers $ \rd -> do
             ptr <-
-                case Alonzo.rdptr (Alonzo.body conwayTx) (toScriptPurpose rd) of
+                case Alonzo.rdptr (Write.txBody (recentEra @era) ledgerTx) (toScriptPurpose rd) of
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: Hi @Anviking!

Unfortunately this line is too long.
Luckily, we can fix it by extracting out the body of `forM` into a separate function:

See: 856af043a64f6fc24fc621909aa435573fd82179

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 2]
```diff
@@ -1124,37 +1048,22 @@ assignScriptRedeemers pparams timeTranslation utxo redeemers tx =
...
-            ]
-
-    addScriptIntegrityHashConway
-        :: era ~ Cardano.ConwayEra => ConwayTx -> ConwayTx
-    addScriptIntegrityHashConway conwayTx =
-        conwayTx & bodyTxL . scriptIntegrityHashTxBodyL .~
+    addScriptIntegrityHash
+        :: Write.RecentEraLedgerConstraints (Write.ShelleyLedgerEra era)
+        => (Write.Tx (Write.ShelleyLedgerEra era))
+        -> (Write.Tx (Write.ShelleyLedgerEra era))
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]:
```suggestion
        => Write.Tx (Write.ShelleyLedgerEra era)
        -> Write.Tx (Write.ShelleyLedgerEra era)
```

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 3]
```diff
@@ -986,36 +966,17 @@ assignScriptRedeemers pparams timeTranslation utxo redeemers tx =
...
-                    { Alonzo.txrdmrs =
-                        Alonzo.Redeemers (Map.fromList nullRedeemers)
-                    }
-                }
+    assignNullRedeemers
+        :: Write.RecentEraLedgerConstraints (Write.ShelleyLedgerEra era)
+        => Write.Tx (Write.ShelleyLedgerEra era)
+        -> Either ErrAssignRedeemers
+            ( Map Alonzo.RdmrPtr Redeemer
+            , (Write.Tx (Write.ShelleyLedgerEra era))
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]:
```suggestion
            , Write.Tx (Write.ShelleyLedgerEra era)
```

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 4]
```diff
@@ -1084,37 +1022,23 @@ assignScriptRedeemers pparams timeTranslation utxo redeemers tx =
...
-        :: Map Alonzo.RdmrPtr (Either ErrAssignRedeemers Alonzo.ExUnits)
-        -> ConwayTx
-        -> Either ErrAssignRedeemers ConwayTx
-    assignExecutionUnitsConway exUnits conwayTx = do
-        let wits = Alonzo.wits conwayTx
-        let Alonzo.Redeemers rdmrs = Alonzo.txrdmrs wits
+    assignExecutionUnits
+        :: Write.RecentEraLedgerConstraints (Write.ShelleyLedgerEra era)
+        => Map Alonzo.RdmrPtr (Either ErrAssignRedeemers Alonzo.ExUnits)
+        -> (Write.Tx (Write.ShelleyLedgerEra era))
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]:
```suggestion
        -> Write.Tx (Write.ShelleyLedgerEra era)
```

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 1 above)
Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: 👍 thanks!

**Review by jonathanknowles — APPROVED** [review]

**Pull request merged**